### ISSUE TITLE
Evaluate replacing vendored LinearAttention with native PyTorch Attention for AptaDiff

### ISSUE THREAD

**Issue by aditi-dsi** [gc-os-ai/pyaptamer#750]
We are currently vendoring [LinearAttention](https://github.com/lucidrains/linear-attention-transformer) and [LocalAttention](https://github.com/lucidrains/local-attention) modules for implementing the transformer layer of AptaDiff (WIP in #739) as done in the [original](https://github.com/wz-create/AptaDiff/blob/master/diffusion/layers/transformer.py) implementation. 

During our discussion, we noted that Linear attention was designed to solve memory bottlenecks for large sequences (e.g., N > 1024) and that our aptamer seqs are shorter with N ~ 40 (attention matrix 40x40 = 1600). This prompts a reevaluation of whether the vendored frameworks are causing unnecessary complexity and can be safely replaced with pytorch's built-in methods.

#### What the Vendored code currently provides
- `O(Nd^2)` Complexity by computing the `(K.V)` matrix first to avoid the full attention matrix. (~ `O(N)` for `N >> d`)
- Explicitly splits attention heads between global linear attention and local sliding-window attention within the same layer (a feature not natively supported out-of-the-box in PyTorch).

#### Trade-offs of migrating to PyTorch native

**Pros:**
*   Native PyTorch can outperform custom linear approximations for shorter sequences. (LinearAttention effects come when `N >> d`)
*   We can drop several lines of tech debt and rely on standard library methods.

**Cons:**
*   Linear attention approximates the attention matrix. Using standard softmax attention changes the math, meaning our forward pass outputs will no longer exactly match the original authors' mathematical formulation.
*   The gradients will flow differently through standard softmax versus a linear approximation, breaking the 1:1 replication.

> Note: The original AptaDiff paper evaluates aptamers with sequence [lengths of 30-40 nt](https://academic.oup.com/bib/article/25/6/bbae517/7828722?login=false#:~:text=The%20target%20sequence,nt%2C%20respectively.), which strongly suggests that if they used linear attention, it may not be for the typical long-sequence motivation (since N=40 is well within the efficient range), hence, we are also not sure about the original motivation behind LinearAttention usage.

We need to align on the core objective for the AptaDiff integration in pyaptamer:
*   **Option A:** If we need **strict reproducibility** and baseline parity with the original paper, then we may need to keep the vendored linear implementation despite the overhead.
*   **Option B:** If we prioritize **architectural optimization** for training from scratch on our own 40-nt datasets, we may transition to native PyTorch attention to maximize speed and maintainability.

**Comment by aditi-dsi** [comment]
cc @siddharth7113

**Comment by siddharth7113** [comment]
> Option B: If we prioritize architectural optimization for training from scratch on our own 40-nt datasets, we may transition to native PyTorch attention to maximize speed and maintainability.

On a side note, here we are not using any pre-trained models, right it would each time trained on a new selex data, for a particular target right?

**Comment by siddharth7113** [comment]
I would put this as an agenda point for community meet, I think If they haven't mentioned why they chose "linear-attention" even in paper, then we could swtch to native attention with mention in docstrings/comments how we wary from them.

**Comment by aditi-dsi** [comment]
> > Option B: If we prioritize architectural optimization for training from scratch on our own 40-nt datasets, we may transition to native PyTorch attention to maximize speed and maintainability.
> 
> On a side note, here we are not using any pre-trained models, right it would each time trained on a new selex data, for a particular target right?

yes, correct.

**Comment by siddharth7113** [comment]
@aditi-dsi What was the decision on this?

**Comment by aditi-dsi** [comment]
@siddharth7113, it was decided that we should introduce `transformer_type` as public parameter (`linear` vs `native` for attention type). In other words, provide dual backend support for AptaDiff transformer layer.
